### PR TITLE
♻️ Refactor: optimization

### LINKED ISSUE #1003: 🐛 group_tolerance/clp_link_tolerance appears to be incorrectly applied
### Version information

- pyglotaran version: main

### Describe the bug

The group_tolerance/clp_link_tolerance appears to not be applied for the first dataset (at least in case there is more than a single dataset), and for non-overlapping datasets the behavior is, unpredictable (possible for the same reason).  

### To Reproduce

1. Run an example with 1 dataset, with the clp_link_tolerance set to a value (much) larger than the spacing between 2 element on the global_axis.

2. Run an example with at least 2 dataset, with the clp_link_tolerance set to a value a 1.01x larger than the spacing between 2 element on the global_axis. Additionally, make sure that one datasets contains 1 or more additional element on the global_axis.

In the specific example showing the issues the datasets have global_axis:
dataset1: 401.025543, 402.901245, 404.941254, 407.021240, ..., 796.960022, 798.969971 (len=159)
dataset2: 400.997772, 402.833344, 404.940002, 407.091125, ..., 796.099976, 798.049988 (len=160)
and with clp_link_tolerance=1.9 this gives problem (also with any other values, but different errors)

To test I would simulate data with a fixed offset between the global_axis plus a small random offset, e.g.
`np.linspace(400,800,51)+(4*np.random.random_sample((1,51))-2)`
`np.linspace(404,800,51)+(4*np.random.random_sample((1,51))-2)`
with `clp_link_tolerance=3.9`
which ensures a random grouping across the board.

### Error messages
A non-exaustive list of error messages
```python
ValueError: conflicting sizes for dimension 'spectral': length <some_number_larger_than_global_axis_1> on 'matrix' and length <length_global_axis_1> on {'time': 
'time', 'spectral': 'spectral', 'left_singular_value_index': 'data_left_singular_vectors', 'singular_value_index': 'data_singular_values', 'right_singular_value_index': 'data_right_singular_vectors'}
```
```python
    full_index_clp_labels = full_clp_labels[i + offset]
IndexError: list index out of range
```
```python
>
    mask = [full_index_clp_labels.index(clp_label) for clp_label in clp_labels]
ValueError: 'c1' is not in list
```
### Expected behavior

If `clp_link_tolerance` is set, it should affect grouping across all datasets. 
If the two dataset do not have equidistant global_axis, and/or an offsetted axis, pyglotaran should not crash

### Additional context

Tracebacks

```python
  File "NTDCK_CK_694_40_20uWQA.py", line 65, in 
main
    result = optimize(scheme)
  File "c:\src\pyglotaran\glotaran\analysis\optimize.py", line 76, in optimize
    return _create_result(
  File "c:\src\pyglotaran\glotaran\analysis\optimize.py", line 137, in _create_result
    group.create_result_data(parameter_history, success=success, add_svd=scheme.add_svd)
  File "c:\src\pyglotaran\glotaran\analysis\optimization_group.py", line 327, in create_result_data
    self._calculator.prepare_result_creation()
  File "c:\src\pyglotaran\glotaran\analysis\optimization_group_calculator_linked.py", line 420, in prepare_result_creation
    full_index_clp_labels = full_clp_labels[i + offset]
IndexError: list index out of range
```

```python
NTDCK_CK_694_40_20uWQA.py", line 65, in 
main
    result = optimize(scheme)
  File "c:\src\pyglotaran\glotaran\analysis\optimize.py", line 76, in optimize
    return _create_result(
  File "c:\src\pyglotaran\glotaran\analysis\optimize.py", line 137, in _create_result
    group.create_result_data(parameter_history, success=success, add_svd=scheme.add_svd)
  File "c:\src\pyglotaran\glotaran\analysis\optimization_group.py", line 330, in create_result_data
    result_data[label] = self.create_result_dataset(label, copy=copy)
  File "c:\src\pyglotaran\glotaran\analysis\optimization_group.py", line 345, in create_result_dataset      
    dataset = self._calculator.create_index_dependent_result_dataset(label, dataset)
  File "c:\src\pyglotaran\glotaran\analysis\optimization_group_calculator_linked.py", line 447, in create_index_dependent_result_dataset
    dataset["matrix"] = (
  File "C:\Anaconda3\envs\py39gta\lib\site-packages\xarray\core\dataset.py", line 1563, in __setitem__      
    self.update({key: value})
  File "C:\Anaconda3\envs\py39gta\lib\site-packages\xarray\core\dataset.py", line 4208, in update
    merge_result = dataset_update_method(self, other)
  File "C:\Anaconda3\envs\py39gta\lib\site-packages\xarray\core\merge.py", line 984, in dataset_update_method
    return merge_core(
  File "C:\Anaconda3\envs\py39gta\lib\site-packages\xarray\core\merge.py", line 640, in merge_core
    dims = calculate_dimensions(variables)
  File "C:\Anaconda3\envs\py39gta\lib\site-packages\xarray\core\dataset.py", line 208, in calculate_dimensions
    raise ValueError(
ValueError: conflicting sizes for dimension 'spectral': length 234 on 'matrix' and length 159 on {'time': 'time', 'spectral': 'spectral', 'left_singular_value_index': 'data_left_singular_vectors', 'singular_value_index': 'data_singular_values', 'right_singular_value_index': 'data_right_singular_vectors'}
```

```python
File "NTDCK_CK_694_40_20uWQA.py", line 65, in 
main
    result = optimize(scheme)
  File "c:\src\pyglotaran\glotaran\analysis\optimize.py", line 76, in optimize
    return _create_result(
  File "c:\src\pyglotaran\glotaran\analysis\optimize.py", line 137, in _create_result
    group.create_result_data(parameter_history, success=success, add_svd=scheme.add_svd)
  File "c:\src\pyglotaran\glotaran\analysis\optimization_group.py", line 327, in create_result_data
    self._calculator.prepare_result_creation()
  File "c:\src\pyglotaran\glotaran\analysis\optimization_group_calculator_linked.py", line 422, in prepare_result_creation
    mask = [full_index_clp_labels.index(clp_label) for clp_label in clp_labels]
  File "c:\src\pyglotaran\glotaran\analysis\optimization_group_calculator_linked.py", line 422, in <listcomp>
    mask = [full_index_clp_labels.index(clp_label) for clp_label in clp_labels]
ValueError: 'c1' is not in list
```

### PR DESCRIPTION
**Pull request by joernweissenborn**
This completely overhauls the the package `glotaran.optimization`

### Change summary

- Removed `OptimizationGroupCalculator`
- created `DataProvider` to provide data for optimization
- created `MatrixProvider` to provide matrix caculations
- created `EstimationProvider` to provide clp and residuals
- created `Optimizer` class
- changed data alignment algorithm 
- added alignment methods `nearest`, `forward` and `backward`

### TODO

- include full models
- add mypy checks
- add docstrings

### Checklist

- [x] ✔️ Passing the tests (mandatory for all PR's)
- [x] 🚧 Added changes to changelog (mandatory for all PR's)
- [x] 👌 Closes issue (mandatory for ✨ feature and 🩹 bug fix PR's)
- [x] 🧪 Adds new tests for the feature (mandatory for ✨ feature and 🩹 bug fix PR's)
- [ ] 📚 Adds documentation of the feature

### Closes issues

closes #1003

### CODE REVIEW

**Comment by s-weigand** [comment]
@joernweissenborn As discussed in slack and on the phone have a look at [🩹⌨️ Remove usage of 'Number' for typing](https://github.com/glotaran/pyglotaran/pull/1060/commits/ab97046a31e7469fbb43eb6a66c421a2b877ca63) and rename `index` to `global_index_value` where appropriate (or a different name when it isn't an int index)

File: glotaran/optimization/matrix_provider.py [hunk 1]
```diff
@@ -0,0 +1,707 @@
...
+
+        Returns
+        -------
+        np.typing.ArrayLike
+            The weighted matrix.
+        """
+        matrix = matrix.copy()
+        for i in nb.prange(matrix.shape[1]):
+            matrix[:, i] *= weight
+        return matrix
```
Comment (s-weigand) [inline glotaran/optimization/matrix_provider.py:50]: This is simple enough that we can use a pure numpy implementation which is actually faster

```python
In [1]: import numba as nb

In [2]: import numpy as np

In [3]: @nb.jit(nopython=True, parallel=True)
   ...: def apply_weight_nb(matrix, weight):
   ...:     matrix = matrix.copy()
   ...:     for i in nb.prange(matrix.shape[1]):
   ...:         matrix[:, i] *= weight
   ...:     return matrix
   ...:

In [4]: def apply_weight_np(matrix, weight):
   ...:     return (matrix.T * weight).T
   ...:

In [5]: matrix = np.ones((1000,200))

In [6]: weights = np.arange(1000)

In [7]: (apply_weight_nb(matrix,weights) == apply_weight_np(matrix,weights)).all()
Out[7]: True

In [8]: %timeit apply_weight_nb(matrix,weights)
749 µs ± 158 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)

In [9]: %timeit apply_weight_np(matrix,weights)
134 µs ± 1.35 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [10]: %timeit apply_weight_nb(matrix,weights)
502 µs ± 20.5 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

File: glotaran/optimization/test/test_constraints.py [hunk 2]
```diff
@@ -17,7 +17,7 @@ def test_constraint(index_dependent, link_clp):
     model.megacomplex["m1"].is_index_dependent = index_dependent
     model.clp_constraints.append(ZeroConstraint.from_dict({"target": "s2"}))
 
-    print("link_clp", link_clp, "index_dependent", index_dependent)
+    print("link_clp", link_clp, "index_dependent", index_dependent)  # T201
```
Comment (s-weigand) [inline glotaran/optimization/test/test_constraints.py:20]: This is an artifact of a badly written noqa and a using `yesqa` see [this commit](https://github.com/glotaran/pyglotaran/commit/3c21f46d8c7a6ecfe5058cc2cc094cef5dd55cf2)

File: setup.cfg [hunk 3]
```diff
@@ -110,6 +110,15 @@ ignore_errors = False
...
 
+[mypy-glotaran.optimization.*]
+ignore_errors = False
+
+[mypy-glotaran.optimization.test.*]
+ignore_errors = True
+
+[mypy-glotaran.project.generators.test.*]
+ignore_errors = True
+
```
Comment (s-weigand) [inline setup.cfg:121]: This should be reactivated see [commit](https://github.com/glotaran/pyglotaran/commit/7b1d50725a0c7b31de04d018ab1654a496cde4c6)

**Review by s-weigand — CHANGES_REQUESTED** [review]
@joernweissenborn Thanks for all the work you put into refactoring the optimization package to make it more readable.

As usual, I got my nitpicks, sorry for that 😅

One of my biggest issues is that classes set attributes outside of the init method which makes it harder to read and also might cause errors if methods are not called in the proper order because the attribute would not exist in that case.

Personally, I think a functional approach makes things easier to unittest and is less error prawn than changing a mutable state, but ofc this also requires functions to take more arguments. 🤷‍♀️

**Comment by s-weigand** [comment]
@joernweissenborn One more questing about typing, why do you use `np.typing.ArrayLike` instead of `np.ndarray`? Are the data a mix of `np.ndarray` and `xr.DataArray`?

File: glotaran/model/dataset_group.py [hunk 4]
```diff
@@ -22,5 +29,36 @@ class DatasetGroupModel:
...
+        ]
+        model_dimensions = {d.get_model_dimension() for d in dataset_models}
+        if len(model_dimensions) != 1:
+            return False
+        global_dimensions = []
+        for dataset in data.values():
+            global_dimensions += [
+                dim for dim in dataset.data.coords if dim not in model_dimensions
+            ]
+        return len(set(global_dimensions)) == 1
```
[the remark this answers is not in the document]
Comment (joernweissenborn) [reply glotaran/model/dataset_group.py:64]: done

**Comment by joernweissenborn** [comment]
> @joernweissenborn One more questing about typing, why do you use `np.typing.ArrayLike` instead of `np.ndarray`? Are the data a mix of `np.ndarray` and `xr.DataArray`?

I like to keep the interface abstract so we later might use e.g. dask or gpu containers.

File: glotaran/optimization/matrix_provider.py [hunk 5]
```diff
@@ -0,0 +1,707 @@
+"""Module containing the matrix provider classes."""
```
[the remark this answers is not in the document]
Comment (joernweissenborn) [reply glotaran/optimization/matrix_provider.py:1]: Sometimes this has to be done this way. But this classes will always just do init -> claculate -> get_result. Many of the steps are not easily unit testable anyway.

**Review by s-weigand — APPROVED** [review]
Looks good now, thnx 😄

**Comment by jsnel** [comment]
I looked in more detail at the remaining discrepancy in the results for the `simultaneous_analysis_6d_disp` integration test and found that the results (estimated parameters) in this PR are *closer* to the true values (values simulated and estimated in TIM as provided by @ism200), so we consider this a regression for the better and accept the change. This mean we will have to re-create the gold standard from the code at the point of this merged PR. 

Spreadsheet with evidence attached: 
[optimized_parameters-6d-compare-PR1118-main-to-PR1060-to-TIM.xlsx](https://github.com/glotaran/pyglotaran/files/9485627/optimized_parameters-6d-compare-PR1118-main-to-PR1060-to-TIM.xlsx)

**Review by jsnel — DISMISSED** [review]
Reviewed ok! See comments for in depth investigation on changes in integration test results.

**Comment by s-weigand** [comment]
After [updating the examples to raise the `maximum_number_function_evaluations` for the `ex_two_datasets` from 15 to 18](https://github.com/glotaran/pyglotaran-examples/pull/72) we only have failing comparison tests for `simultaneous_analysis_6d_disp` as [explained by @jsnel ](https://github.com/glotaran/pyglotaran/pull/1060#issuecomment-1236428361)
See [CI run 1901](https://github.com/glotaran/pyglotaran/runs/8179855527?check_suite_focus=true) or attached logs 
[PR1060-CI-run-1901.zip](https://github.com/glotaran/pyglotaran/files/9485635/PR1060-CI-run-1901.zip)

We will now update the 'gold standard' and run the CI again

**Review by jsnel — DISMISSED** [review]

**Review by jsnel — APPROVED** [review]

**Pull request merged**